### PR TITLE
[POC] New signature v2

### PR DESCRIPTION
**Pull request by rullzer**
POC of https://github.com/nextcloud/server/issues/18464

TODO:
* cleanup code (better v1/v2 handling separation)
* fixup tests

### CODE REVIEW

File: lib/private/IntegrityCheck/Checker.php [hunk 1]
```diff
@@ -275,6 +272,34 @@ public function writeAppSignature($path,
...
+
+		try {
+			$xml = new \SimpleXMLElement($content);
+			$minNextcloudVersion = $xml->dependencies->nextcloud['min-version'];
+
+			if ($minNextcloudVersion === NULL) {
+				return 1;
+			}
+
+			if ((int)$minNextcloudVersion >= 19) {
```
Comment (ChristophWurst) [inline lib/private/IntegrityCheck/Checker.php]: it's possible to specify a version like `19.0.2`, then the int cast will fail

File: lib/private/IntegrityCheck/Checker.php [hunk 2]
```diff
@@ -357,10 +380,27 @@ private function verify(string $signaturePath, string $basePath, string $certifi
 		$rsa->setMGFHash('sha512');
 		// See https://tools.ietf.org/html/rfc3447#page-38
 		$rsa->setSaltLength(0);
-		if(!$rsa->verify(json_encode($expectedHashes), $signature)) {
+
+		$version = 1;
+		if (isset($signatureData['version'])) {
+			$version = $signatureData['version'];
+		}
```
Comment (ChristophWurst) [inline lib/private/IntegrityCheck/Checker.php]: or ``$version = $signatureData['version'] ?? 1`` for short :)

Comment (rullzer) [reply lib/private/IntegrityCheck/Checker.php]: fair enough.

File: lib/private/IntegrityCheck/Checker.php (hunk 1 above)
Comment (rullzer) [reply lib/private/IntegrityCheck/Checker.php]: o right. Yeah then lets use some version parsing thing

File: lib/private/IntegrityCheck/Checker.php [hunk 3]
```diff
@@ -229,19 +215,28 @@ private function generateHashes(\RecursiveIteratorIterator $iterator,
...
+		if ($version === 2) {
+			$hashesOutput = $encodedHashed;
+		}
 
 		return [
-				'hashes' => $hashes,
+				'hashes' => $hashesOutput,
 				'signature' => base64_encode($signature),
 				'certificate' => $certificate->saveX509($certificate->currentCert),
+				'version' => 2,
```
Comment (ChristophWurst) [inline lib/private/IntegrityCheck/Checker.php]:
```suggestion
				'version' => $version,
```

File: lib/private/IntegrityCheck/Checker.php [hunk 4]
```diff
@@ -275,6 +272,34 @@ public function writeAppSignature($path,
 		}
 	}
 
+	/**
+	 * Checkl if we can sign the app with the v2 signature
```
Comment (splitt3r) [inline lib/private/IntegrityCheck/Checker.php]:
```suggestion
	 * Check if we can sign the app with the v2 signature
```

**Pull request closed without merging**